system/lib/websocket/websocket_to_posix_socket.cpp

```diff
@@ -0,0 +1,965 @@
+  } d;
+
+  PosixSocketCallResult *b = allocate_call_result(sizeof(SocketCallResultHeader));
+  d.header.callId = b->callId;
+  d.header.function = POSIX_SOCKET_MSG_ACCEPT;
+  d.socket = socket;
+  d.address_len = address_len ? *address_len : 0;
+  emscripten_websocket_send_binary(bridgeSocket, &d, sizeof(d));
+
+  struct Result {
+    SocketCallResultHeader header;
+    int address_len;
+    uint8_t address[];
+  };
+
+  wait_for_call_result(b);
+  int ret = b->data->ret;
+  if (ret == 0)
+  {
+    Result *r = (Result*)b->data;
+    int realAddressLen = MIN(b->bytes - sizeof(Result), r->address_len);
+    int copiedAddressLen = MIN(*address_len, realAddressLen);
+    if (address) memcpy(address, r->address, copiedAddressLen);
+    if (address_len) *address_len = realAddressLen;
+  }
+  else
+  {
+    errno = b->data->errno_;
+  }
+  free_call_result(b);
+  return ret;
+}
+
+int getsockname(int socket, struct sockaddr *address, socklen_t *address_len)
+{
+#ifdef POSIX_SOCKET_DEBUG
+  emscripten_log(EM_LOG_NO_PATHS | EM_LOG_CONSOLE | EM_LOG_ERROR | EM_LOG_JS_STACK, "getsockname(socket=%d,address=%p,address_len=%p)\n", socket, address, address_len);
+#endif
+
+  struct {
+    SocketCallHeader header;
+    int socket;
+    uint32_t/*socklen_t*/ address_len;
+  } d;
+
+  struct Result {
+    SocketCallResultHeader header;
+    int address_len;
+    uint8_t address[];
+  };
+
+  PosixSocketCallResult *b = allocate_call_result(sizeof(Result));
+  d.header.callId = b->callId;
+  d.header.function = POSIX_SOCKET_MSG_GETSOCKNAME;
+  d.socket = socket;
+  d.address_len = *address_len;
+  emscripten_websocket_send_binary(bridgeSocket, &d, sizeof(d) + *address_len - MAX_SOCKADDR_SIZE);
+
+  wait_for_call_result(b);
+  int ret = b->data->ret;
+  if (ret == 0)
+  {
+    Result *r = (Result*)b->data;
+    int realAddressLen = MIN(b->bytes - sizeof(Result), r->address_len);
+    int copiedAddressLen = MIN(*address_len, realAddressLen);
+    if (address) memcpy(address, r->address, copiedAddressLen);
+    if (address_len) *address_len = realAddressLen;
+  }
+  else
+  {
+    errno = b->data->errno_;
+  }
+  free_call_result(b);
+  return ret;
+}
+
+int getpeername(int socket, struct sockaddr *address, socklen_t *address_len)
+{
+#ifdef POSIX_SOCKET_DEBUG
+  emscripten_log(EM_LOG_NO_PATHS | EM_LOG_CONSOLE | EM_LOG_ERROR | EM_LOG_JS_STACK, "getpeername(socket=%d,address=%p,address_len=%p)\n", socket, address, address_len);
+#endif
+
+  struct {
+    SocketCallHeader header;
+    int socket;
+    uint32_t/*socklen_t*/ address_len;
+  } d;
+
+  struct Result {
+    SocketCallResultHeader header;
+    int address_len;
+    uint8_t address[];
+  };
+
+  PosixSocketCallResult *b = allocate_call_result(sizeof(Result));
+  d.header.callId = b->callId;
+  d.header.function = POSIX_SOCKET_MSG_GETPEERNAME;
+  d.socket = socket;
+  d.address_len = *address_len;
+  emscripten_websocket_send_binary(bridgeSocket, &d, sizeof(d) + *address_len - MAX_SOCKADDR_SIZE);
```

Comment:
```suggestion
  emscripten_websocket_send_binary(bridgeSocket, &d, sizeof(d);
```

I'm not sure why the size passed to `emscripten_websocket_send_binary` has `address_len` added and `MAX_SOCKADDR_SIZE` subtracted, but removing that fixes the issue I described in this comment. https://github.com/emscripten-core/emscripten/pull/7672#issuecomment-470742919